README.md
---
title: Light Statusbar
description: Plugin for Cordova (or PhoneGap) 3.0+ to enable light status bar on Android 6.0+ or MIUI6+.
---
<!---
# license: Licensed to the Apache Software Foundation (ASF) under one
#         or more contributor license agreements.  See the NOTICE file
#         distributed with this work for additional information
#         regarding copyright ownership.  The ASF licenses this file
#         to you under the Apache License, Version 2.0 (the
#         "License"); you may not use this file except in compliance
#         with the License.  You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#         Unless required by applicable law or agreed to in writing,
#         software distributed under the License is distributed on an
#         "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#         KIND, either express or implied.  See the License for the
#         specific language governing permissions and limitations
#         under the License.
-->

# cordova-plugin-light-statusbar

LightStatusbar
======

> The `LightStatusbar` object can enable/disable light status bar on Android 6.0+ or MIUI6+.

## Installation

    cordova plugin add https://github.com/michaelins/cordova-plugin-light-statusbar.git

Methods
-------
This plugin defines global `LightStatusBar` object.

Although in the global scope, it is not available until after the `deviceready` event.

    document.addEventListener("deviceready", onDeviceReady, false);
    function onDeviceReady() {
        console.log(LightStatusBar);
    }

- LightStatusBar.isSupported(successCallback)
- LightStatusBar.setStatusBarColor(hexString)

LightStatusBar.isSupported
=================

    LightStatusBar.isSupported(function(result){
      if(result=="true"){
        
      } else {
   
      }
    });

Description
-----------

check if current platform supports light background status bar or not.


Supported Platforms
-------------------

- Android

LightStatusBar.setStatusBarColor
=================

    LightStatusBar.setStatusBarColor(hexString);

Description
-----------

set the status background color.


Supported Platforms
-------------------

- Android

Quick Example
-------------

    LightStatusBar.setStatusBarColor("#FFFFFF");

RELEASENOTES.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->
# Release Notes

### 1.0.0 (Oct 21, 2016) (First release as a Cordova Plugin)
* Initial commit


www/lightstatusbar.js
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/

/* global cordova */

var exec = require('cordova/exec');

var LightStatusBar = {

    isSupported: function (success) {
      exec(success, null, "LightStatusBar", "isSupported", []);
    },

    setStatusBarColor: function (hexString) {
      if (hexString.charAt(0) !== "#") {
        hexString = "#" + hexString;
      }

      if (hexString.length === 4) {
        var split = hexString.split("");
        hexString = "#" + split[1] + split[1] + split[2] + split[2] + split[3] + split[3];
      }

      exec(null, null, "LightStatusBar", "setStatusBarColor", [hexString]);
    },

    enable: function () {
        // dark text ( to be used on a light background )
        exec(null, null, "LightStatusBar", "enable", []);
    },

    disable: function () {
        // light text ( to be used on a dark background )
        exec(null, null, "LightStatusBar", "disable", []);
    }

};

module.exports = LightStatusBar;


src/android/StatusBar.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/
package org.apache.cordova.statusbar;

import android.app.Activity;
import android.graphics.Color;
import android.os.Build;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.LOG;
import org.apache.cordova.PluginResult;
import org.json.JSONException;

public class StatusBar extends CordovaPlugin {
    private static final String TAG = "StatusBar";

    /**
     * Sets the context of the Command. This can then be used to do things like
     * get file paths associated with the Activity.
     *
     * @param cordova The context of the main Activity.
     * @param webView The CordovaWebView Cordova is running in.
     */
    @Override
    public void initialize(final CordovaInterface cordova, CordovaWebView webView) {
        LOG.v(TAG, "StatusBar: initialization");
        super.initialize(cordova, webView);

        this.cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                // Clear flag FLAG_FORCE_NOT_FULLSCREEN which is set initially
                // by the Cordova.
                Window window = cordova.getActivity().getWindow();
                window.clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
            }
        });
    }

    /**
     * Executes the request and returns PluginResult.
     *
     * @param action            The action to execute.
     * @param args              JSONArry of arguments for the plugin.
     * @param callbackContext   The callback id used when calling back into JavaScript.
     * @return                  True if the action was valid, false otherwise.
     */
    @Override
    public boolean execute(final String action, final CordovaArgs args, final CallbackContext callbackContext) throws JSONException {
        LOG.v(TAG, "Executing action: " + action);
        final Activity activity = this.cordova.getActivity();
        final Window window = activity.getWindow();

        if ("_ready".equals(action)) {
            boolean statusBarVisible = (window.getAttributes().flags & WindowManager.LayoutParams.FLAG_FULLSCREEN) == 0;
            callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, statusBarVisible));
            return true;
        }

        if ("show".equals(action)) {
            this.cordova.getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    // SYSTEM_UI_FLAG_FULLSCREEN is available since JellyBean, but we
                    // use KitKat here to be aligned with "Fullscreen"  preference
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        int uiOptions = window.getDecorView().getSystemUiVisibility();
                        uiOptions &= ~View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN;
                        uiOptions &= ~View.SYSTEM_UI_FLAG_FULLSCREEN;

                        window.getDecorView().setSystemUiVisibility(uiOptions);
                    }

                    // CB-11197 We still need to update LayoutParams to force status bar
                    // to be hidden when entering e.g. text fields
                    window.clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                }
            });
            return true;
        }

        if ("hide".equals(action)) {
            this.cordova.getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    // SYSTEM_UI_FLAG_FULLSCREEN is available since JellyBean, but we
                    // use KitKat here to be aligned with "Fullscreen"  preference
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        int uiOptions = window.getDecorView().getSystemUiVisibility()
                                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                                | View.SYSTEM_UI_FLAG_FULLSCREEN;

                        window.getDecorView().setSystemUiVisibility(uiOptions);
                    }

                    // CB-11197 We still need to update LayoutParams to force status bar
                    // to be hidden when entering e.g. text fields
                    window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                }
            });
            return true;
        }

        if ("backgroundColorByHexString".equals(action)) {
            this.cordova.getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        setStatusBarBackgroundColor(args.getString(0));
                    } catch (JSONException ignore) {
                        LOG.e(TAG, "Invalid hexString argument, use f.i. '#777777'");
                    }
                }
            });
            return true;
        }

        return false;
    }

    private void setStatusBarBackgroundColor(final String colorPref) {
        if (Build.VERSION.SDK_INT >= 21) {
            if (colorPref != null && !colorPref.isEmpty()) {
                final Window window = cordova.getActivity().getWindow();
                // Method and constants not available on all SDKs but we want to be able to compile this code with any SDK
                window.clearFlags(0x04000000); // SDK 19: WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
                window.addFlags(0x80000000); // SDK 21: WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                try {
                    // Use light status bar
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        int uiOptions = window.getDecorView().getSystemUiVisibility() | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                        window.getDecorView().setSystemUiVisibility(uiOptions);
                    }
                    // Using reflection makes sure any 5.0+ device will work without having to compile with SDK level 21
                    window.getClass().getDeclaredMethod("setStatusBarColor", int.class).invoke(window, Color.parseColor(colorPref));
                } catch (IllegalArgumentException ignore) {
                    LOG.e(TAG, "Invalid hexString argument, use f.i. '#999999'");
                } catch (Exception ignore) {
                    // this should not happen, only in case Android removes this method in a version > 21
                    LOG.w(TAG, "Method window.setStatusBarColor not found for SDK level " + Build.VERSION.SDK_INT);
                }
            }
        }
    }
}


src/com/github/michaelins/lightstatusbar/LightStatusBar.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/
package com.github.michaelins.lightstatusbar;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.LOG;
import org.json.JSONException;

import com.readystatesoftware.systembartint.SystemBarTintManager;
import android.annotation.TargetApi;
import android.app.Activity;
import android.graphics.Color;
import android.os.Build;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

public class LightStatusBar extends CordovaPlugin {

	private static final String TAG = "LightStatusBar";

	/**
	 * Executes the request and returns PluginResult.
	 *
	 * @param action
	 *            The action to execute.
	 * @param args
	 *            JSONArry of arguments for the plugin.
	 * @param callbackContext
	 *            The callback id used when calling back into JavaScript.
	 * @return True if the action was valid, false otherwise.
	 */
	@Override
	public boolean execute(final String action, final CordovaArgs args, final CallbackContext callbackContext)
			throws JSONException {

		LOG.v(TAG, "Executing action: " + action);
		final Activity activity = this.cordova.getActivity();
		final Window window = activity.getWindow();

		if ("isSupported".equals(action)) {
			this.cordova.getActivity().runOnUiThread(new Runnable() {
				@Override
				public void run() {
					if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M || SystemBarTintManager.IsMiuiV6Plus()) {
						callbackContext.success("true");
					} else {
						callbackContext.success("false");
					}
				}
			});
			return true;
		}

		if ("setStatusBarColor".equals(action)) {
			this.cordova.getActivity().runOnUiThread(new Runnable() {
				@Override
				public void run() {
					try {
					  webView.getView().setFitsSystemWindows(true);
						if (SystemBarTintManager.IsMiuiV6Plus()) {
							// MIUI6+ light status bar
							if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
								setTranslucentStatus(window, true);
							}
							SystemBarTintManager tintManager = new SystemBarTintManager(activity);
							tintManager.setStatusBarTintEnabled(true);
							tintManager.setStatusBarTintColor(Color.parseColor(args.getString(0)));
							tintManager.setStatusBarDarkMode(true, activity);
						} else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
							// 6.0+ light status bar
							window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
							window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
							int uiOptions = window.getDecorView().getSystemUiVisibility()
									| View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
							window.getDecorView().setSystemUiVisibility(uiOptions);
							setStatusBarBackgroundColor(args.getString(0));
						} else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
							setStatusBarBackgroundColor(args.getString(0));
						}
					} catch (JSONException ignore) {
						LOG.e(TAG, "Invalid hexString argument, use f.i. '#777777'");
					}
				}
			});
			return true;
		}

		if ("enable".equals(action)) {
			this.cordova.getActivity().runOnUiThread(new Runnable() {
				@Override
				public void run() {
				}
			});
			return true;
		}

		if ("disable".equals(action)) {
			this.cordova.getActivity().runOnUiThread(new Runnable() {
				@Override
				public void run() {
				}
			});
			return true;
		}
		return false;
	}

	private void setStatusBarBackgroundColor(final String colorPref) {
		if (colorPref != null && !colorPref.isEmpty()) {
			final Window window = cordova.getActivity().getWindow();
			// Method and constants not available on all SDKs but we want to
			// be able to compile this code with any SDK
			window.clearFlags(0x04000000); // SDK 19:
											// WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
			window.addFlags(0x80000000); // SDK 21:
											// WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
			try {
				// Using reflection makes sure any 5.0+ device will work
				// without having to compile with SDK level 21
				window.getClass().getDeclaredMethod("setStatusBarColor", int.class).invoke(window,
						Color.parseColor(colorPref));
			} catch (IllegalArgumentException ignore) {
				LOG.e(TAG, "Invalid hexString argument, use f.i. '#999999'");
			} catch (Exception ignore) {
				// this should not happen, only in case Android removes this
				// method in a version > 21
				LOG.w(TAG, "Method window.setStatusBarColor not found for SDK level " + Build.VERSION.SDK_INT);
			}
		}
	}

	@TargetApi(Build.VERSION_CODES.KITKAT)
	private void setTranslucentStatus(Window win, boolean on) {
		WindowManager.LayoutParams winParams = win.getAttributes();
		final int bits = WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS;
		if (on) {
			winParams.flags |= bits;
		} else {
			winParams.flags &= ~bits;
		}
		win.setAttributes(winParams);
	}
}
